fda/lfd.py

```diff
@@ -0,0 +1,91 @@
+import numpy as np
+
+from fda.basis import FDataBasis, Constant, _same_domain
+
+__author__ = "Pablo Pérez Manso"
+__email__ = "92manso@gmail.com"
+
+
+class Lfd:
+    """Defines the structure of a linear differential operator function system
+
+    .. math::
+        Lx(t) = b_0(t) x(t) + b_1(t) x'(x) + \dots + b_{n-1}(t) d^{n-1}(x(t)) + b_n(t) d^n(x(t))
+
+    Attributes:
+        nderiv (int): the order of the operator. It's the n coefficient in the equation above.
+
+        bwtlist (list):  A FDataBasis objects list of length either nderiv
+
+    """
+
+    def __init__(self, param=0, domain_range=(0, 1)):
+        """Lfd Constructor
+
+        Args:
+            param (int, list, FDataBasis optional): If the given parameter is an integer, it'll be the order of
+                the highest derivative.
+                  If its a list, of integers or FDataBasis objects, it will be taken as the weight functions
+                  for the derivatives.
+                  If its a FDataBasis object, each sample will be a weight function and nderiv will be the number
+                  of samples
+
+            domain_range (tuple or list of tuples, optional): Definition of the
+                interval where the weight functions are defined. Defaults to (0,1).
+        """
+
+        if isinstance(param, list):
+            if all(isinstance(n, int) for n in param):
+                self.nderiv = len(param)
+                self.bwtlist = (FDataBasis(Constant(domain_range), np.array(param).reshape(-1, 1)).to_list()
+                                if self.nderiv != 0 else [])
+            elif all(isinstance(n, FDataBasis) for n in param):
+                if all([_same_domain(param[0].domain_range, x.domain_range) for x in param]):
+                    self.nderiv = len(param)
+                    self.bwtlist = param
+                else:
+                    raise ValueError("FDataBasis objects in the list has not the same domain_range")
+            else:
+                raise ValueError("The elements of the list are neither integers or FDataBasis objects")
+
+        elif isinstance(param, FDataBasis):
+            self.nderiv = param.nsamples
+            self.bwtlist = param.to_list()
+
+        elif isinstance(param, int):
```

Comment:
Now with the code in front of me I understand better why you chose two parameters instead of one, and I think that I agree with that approach. One would expect that f(i) and f([i]) are not very different.

Thus, it is better that the constructor accepts two parameters, as you said it did before: `coefficients`, and `order`. The second can be deduced from the first one if missing, and the first one would be `(0, 0, 0, ..., 0, 1)` if missing. The user MUST give only one of them.

Sorry for making you redo your work.